### PR TITLE
feat: add Verstraete-Cirac fermion-to-qubit Encoding(#482)

### LINKED ISSUE #482: Add Verstraete-Cirac Fermion-to-Qubit Encoding
**Is your feature request related to a problem? Please describe.**

 QDK/Chemistry currently ships only Jordan-Wigner, Bravyi-Kitaev, parity, and SCBK fermion-to-qubit encodings. For 2D-local fermionic problems -- Fermi-Hubbard, lattice gauge theories, fermionic tensor-network ansatze, real-space-grid chemistry -- every nearest-neighbor hop maps to a long Pauli string under these encodings, which inflates Trotter circuit depth and measurement cost.

 **Describe the solution you'd like**

 A working **Verstraete-Cirac encoding** exposed as a new `MajoranaMapping` factory and consumable through `QubitMapper`, with the resulting `QubitHamiltonian` correctly representing the original fermionic Hamiltonian on the physical subspace. Mapper consumers (energy estimation, state preparation, Trotterization) should be able to use it without special-casing.

 **Describe alternatives you've considered**

 Other redundant encodings (Derby-Klassen compact, Bravyi-Kitaev superfast, Setia-Whitfield) would address the same locality problem with different qubit-overhead and stabilizer-structure trade-offs. Verstraete-Cirac is requested first asthe historically canonical choice; the others are natural follow-ups.

 Circuit-level Pauli grouping helps measurement cost under JW/BK but does not reduce Trotter depth.

 **Acceptance criteria**

 - A new `MajoranaMapping` factory for Verstraete-Cirac accepts a general 2D lattice of at least sizes 2x2, 2x3, 3x3, and 4x4 (single spin species) and returns a mapping that `QubitMapper` consumes without error.
 - For the Fermi-Hubbard model on a 2x2 lattice with open boundaries, t=1, U=4, at half-filling, the four lowest eigenvalues of the resulting `QubitHamiltonian` restricted to the codespace match the Jordan-Wigner result to within `1e-10` absolute error.
 - For Fermi-Hubbard on open LxL lattices with L in {2, 3, 4}, the maximum Pauli weight over all nearest-neighbor hopping terms produced by the mapper is the same finite integer for all three L (i.e. independent of system size).
 - The constructed mapping survives JSON and HDF5 round-trips, and the round-tripped mapping produces a `QubitHamiltonian` whose terms equal the original's term-by-term (Pauli string and coefficient).
 - A new pytest module under `python/tests/` covers all of the above; the full existing test suite continues to pass; `pre-commit` succeeds on the diff.
 - The encoding is listed in `docs/source/user/comprehensive/algorithms/qubit_mapper.rst` under "Supported encodings" in the same format as the other entries, and citations for the three references below are added to`docs/source/references.bib`.

 **Additional context**

 - F. Verstraete and J. I. Cirac, *Mapping local Hamiltonians of fermions to local Hamiltonians of spins*, J. Stat. Mech. (2005) P09012.
 - J. D. Whitfield, V. Havlicek, and M. Troyer, *Local spin operators for fermion simulations*, Phys. Rev. A **94**, 030301(R) (2016) [arXiv:1605.09789].
 - V. Havlicek, M. Troyer, and J. D. Whitfield, *Operator locality in the quantum simulation of fermionic models*, Phys. Rev. A **95**, 032332 (2017) [arXiv:1701.07072].

### PR DESCRIPTION
**Pull request by GoiBasia**
Closes #482 

## Summary

This local fix addresses Microsoft QDK-Chemistry issue [#482](https://github.com/microsoft/qdk-chemistry/issues/482). The implementation adds Verstraete-Cirac as a native C++ `MajoranaMapping` factory, exposed through Python and consumable by the existing `QubitMapper` path. The new factory accepts a `LatticeGraph`, builds the physical and auxiliary Majorana layout, dresses non-path hopping terms with auxiliary stabilizers, and returns a mapping that downstream Hamiltonian construction can use without special-casing.

- Implemented `VC mapping` end-to-end in C++/Python, with `local auxiliary penalties`.
- Added the core `MajoranaMapping::verstraete_cirac(lattice)` implementation.
- Added stabilizer and `auxiliary_penalty_terms` support, including serialization and Python exposure.
- Changed the mapper engine to penalize local auxiliary cycle terms instead of raw nonlocal stabilizers.
- Fixed penalty scaling so two-body Hubbard terms contribute to the stabilizer penalty strength.
- Added `dfs_ordering` and `LatticeGraph::permute(...)` support for locality-aware lattice ordering.
- Added regression coverage for VC dimensions, serialization, commutation, Pauli-weight scaling, penalty scaling, and spectral validation.
- Updated user docs and references for the Verstraete-Cirac encoding.

## Verification

All python and C++ tests pass and the local test coverage focuses on the issue acceptance criteria and the main review risks:

- required lattice sizes such as `2x2`, `2x3`, `3x3`, and `4x4`;
- JSON and HDF5 round-trips of mappings containing stabilizers and auxiliary penalty terms;
- stabilizer commutation and explicit codespace projection;
- Fermi-Hubbard `2x2` spectral agreement at half-filling inside the stabilizer `+1` codespace;
- Pauli-weight locality checks for nearest-neighbor hopping without using the previous magic coefficient cutoff;
- regression coverage that stabilizer penalty scaling includes two-body Hubbard interactions;
- C++ lattice-graph tests for DFS ordering and edge-color preservation.

**Factory + QubitMapper for 2x2, 2x3, 3x3, 4x4 single spin:**
2x2: qh_terms=29, qh_qubits=12, mapping_modes=8
2x3: qh_terms=47, qh_qubits=20, mapping_modes=12
3x3: qh_terms=80, qh_qubits=32, mapping_modes=18
4x4: qh_terms=157, qh_qubits=60, mapping_modes=32

**2x2 Fermi-Hubbard, t=1, U=4, half-filling, VC codespace vs JW eigenvalues:**

- JW first 4: `[-10.102748483462, -9.806423851823, -9.806423851823, -9.806423851823]`
- VC restricted to half-filling + stabilizer codespace: same values
- Max absolute error: `5.329070518200751e-15`, which is below `1e-10`

It restricts JW to half-filling, projects VC into the stabilizer `+1` codespace, verifies particle number/stabilizers, then compares `eigs_vc[:4]` to `eigs_jw[:4]` with `atol=1e-10`.

**Pauli-weight scaling for L in {2,3,4}:**

```text
spin_species 1: L=2 -> 4, L=3 -> 4, L=4 -> 4
spin_species 2: L=2 -> 4, L=3 -> 4, L=4 -> 4
```

The test uses `create_huckel_hamiltonian(..., epsilon=0, t=1)` to isolate the hopping part rather than constructing the full Hubbard Hamiltonian with `U`. For this criterion, that is technically fine because the checked object is only the nearest-neighbor hopping terms; the Hubbard onsite interaction does not affect their Pauli weight.

### CODE REVIEW

File: cpp/src/qdk/chemistry/data/lattice_graph.cpp [hunk 1]
```diff
@@ -947,6 +1068,49 @@ void LatticeGraph::hash_update(qdk::chemistry::utils::HashContext& ctx) const {
...
+  std::uint64_t V = graph.num_sites();
+  const auto& adj = graph.sparse_adjacency_matrix();
+
+  // Reorder the adjacency matrix using Eigen's PermutationMatrix
+  Eigen::PermutationMatrix<Eigen::Dynamic, Eigen::Dynamic, int> P(
+      static_cast<Eigen::Index>(V));
+  for (std::uint64_t i = 0; i < V; ++i) {
+    P.indices()[static_cast<Eigen::Index>(i)] = static_cast<int>(path[i]);
+  }
+  Eigen::SparseMatrix<double> new_adj = P * adj * P.transpose();
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply cpp/src/qdk/chemistry/data/lattice_graph.cpp:1076]: Done(https://github.com/microsoft/qdk-chemistry/pull/536/commits/61c49188dec2c721dc0c2a1a2304f96c1567b831)
 `LatticeGraph::permute` now validates `path` before using it: it checks that the path length matches `graph.num_sites()`, every vertex index is in range, and every vertex appears exactly once. Invalid inputs now throw `std::invalid_argument` before Eigen permutation setup or inverse-map writes, eliminating the out-of-bounds read/write risk.

File: cpp/src/qdk/chemistry/data/lattice_graph.cpp [hunk 2]
```diff
@@ -28,6 +28,106 @@ static void add_edge(std::vector<Triplet>& triplets, int i, int j, double t) {
...
+    }
+  }
+  return {};
+}
+
+LatticeGraph apply_dfs_ordering(const LatticeGraph& graph) {
+  auto path = find_hamiltonian_path(graph.sparse_adjacency_matrix());
+  if (path.empty()) {
+    throw std::runtime_error("No Hamiltonian path found in the lattice graph.");
+  }
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply cpp/src/qdk/chemistry/data/lattice_graph.cpp]: Done(https://github.com/microsoft/qdk-chemistry/pull/536/commits/61c49188dec2c721dc0c2a1a2304f96c1567b831)
`dfs_ordering` no longer runs the exact Hamiltonian-path backtracking search or throws when no Hamiltonian path exists. It now uses a cheap deterministic DFS traversal order with stable neighbor ordering and falls back across disconnected components to always produce a valid permutation. I also updated the tests to validate the new intended behavior: deterministic ordering, preserved topology, preserved edge count/nonzero count, and valid remapped edge coloring, including the 4x4 square-lattice case.

**Comment by GoiBasia** [comment]
This update refines the Verstraete-Cirac implementation for issue #482  including both the default single-species lattice use case and QDK-Chemistry’s spinful electronic-structure convention. The mapper path has been updated accordingly across C++, Python bindings, and the Python `QdkQubitMapper`. The native mapping engine now accepts both `n_spatial` single-species mappings and `2 * n_spatial` spinful mappings, while preserving the existing Verstraete-Cirac auxiliary stabilizer and penalty-term behavior. The tests were also updated to cover both modes. The local coverage now checks single-species factory construction and mapper execution for `2x2`, `2x3`, `3x3`, and `4x4` lattices, explicit spinful mapping with `spin_species=2`, invalid spin-species rejection, Pauli-weight scaling for spinful Hubbard hopping terms, penalty scaling with two-body interactions, and spectral validation for both spinless Hückel and spinful Hubbard models.

File: python/src/pybind11/data/lattice_graph.cpp [hunk 3]
```diff
@@ -295,7 +295,7 @@ With periodic boundary condition:
     >>> ring = LatticeGraph.chain(6, periodic=True)
 )",
                            py::arg("n"), py::arg("periodic") = false,
-                           py::arg("t") = 1.0);
+                           py::arg("t") = 1.0, py::arg("dfs_ordering") = false);
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply python/src/pybind11/data/lattice_graph.cpp:301]: Done(https://github.com/microsoft/qdk-chemistry/pull/536/commits/b7ceb600c9839fb39f2706b64447e40c9f4a0cfb)
I updated `lattice_graph.cpp`for chain, square, triangular, honeycomb, and kagome, adding:
dfs_ordering (bool, optional): If True, relabel sites using deterministic DFS traversal order. Defaults to False.

File: python/src/qdk_chemistry/data/__init__.py [hunk 4]
```diff
@@ -130,6 +130,62 @@
...
+    only exposes ``verstraete_cirac(lattice)`` and returns a two-species mapping.
+    Keep that environment runnable by deriving the alpha-sector bilinear table
+    and stabilizer metadata for the requested single-species mapping.
+    """
+    try:
+        return _native_verstraete_cirac(lattice, spin_species)
+    except TypeError:
+        if spin_species not in {1, 2}:
+            raise ValueError("verstraete-cirac requires spin_species to be either 1 or 2") from None
+
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply python/src/qdk_chemistry/data/__init__.py]: Thanks for pointing this out. The modification to `__init__.py` was included accidentally and has been reverted to the original version. No changes to this file are intended in this PR.

File: python/src/qdk_chemistry/algorithms/qubit_mapper/qdk_qubit_mapper.py [hunk 5]
```diff
@@ -31,6 +32,120 @@
...
+        for q in range(n_spatial):
+            h_pq = float(h1_alpha[p, q])
+            if abs(h_pq) > integral_threshold:
+                accumulate_epq(p, q, h_pq)
+
+    two_body = np.asarray(hamiltonian.get_two_body_integrals()[0]).reshape(-1)
+    expected_size = n_spatial**4
+    if two_body.size == expected_size and np.any(
+        np.abs(two_body) > integral_threshold
+    ):
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply python/src/qdk_chemistry/algorithms/qubit_mapper/qdk_qubit_mapper.py]: Done([b7ceb60](https://github.com/microsoft/qdk-chemistry/pull/536/commits/b7ceb600c9839fb39f2706b64447e40c9f4a0cfb))
I updated `_python_single_species_map_hamiltonian` so it now checks `hamiltonian.has_two_body_integrals()` before calling `get_two_body_integrals()`.

**Comment by wavefunction91** [comment]
Closed in favor of #505.
